overlord/assertstate/assertmgr.go

```diff
@@ -263,3 +265,111 @@ func RefreshSnapDeclarations(s *state.State, userID int) error {
 	}
 	return doFetch(s, userID, fetching)
 }
+
+type refreshControlError struct {
+	errs []error
+}
+
+func (e *refreshControlError) Error() string {
+	l := []string{""}
+	for _, e := range e.errs {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("refresh control errors:%s", strings.Join(l, "\n - "))
+}
+
+// ValidateRefreshes validates the refresh candidate revisions represented by the snapInfos, looking for the needed refresh control validation assertions, it returns a validated subset in validated and a summary error if not all candidates validated.
+func ValidateRefreshes(s *state.State, snapInfos []*snap.Info, userID int) (validated []*snap.Info, err error) {
+	// maps gated snap-ids to gating snap-ids
+	controlled := make(map[string][]string)
+	// maps gating snap-ids to their snap names
+	gatingNames := make(map[string]string)
+
+	db := DB(s)
+	snapStates, err := snapstate.All(s)
+	if err != nil {
+		return nil, err
+	}
+	for snapName, snapState := range snapStates {
+		info, err := snapState.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		if info.SnapID == "" {
+			continue
+		}
+		gatingID := info.SnapID
+		a, err := db.Find(asserts.SnapDeclarationType, map[string]string{
+			"series":  release.Series,
+			"snap-id": gatingID,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("internal error: cannot find snap declaration for installed snap %q (id %q): err", snapName, gatingID)
+		}
+		decl := a.(*asserts.SnapDeclaration)
+		control := decl.RefreshControl()
+		if len(control) == 0 {
+			continue
+		}
+		gatingNames[gatingID] = decl.SnapName()
+		for _, gatedID := range control {
+			controlled[gatedID] = append(controlled[gatedID], gatingID)
+		}
+	}
+
+	var errs []error
+	for _, candInfo := range snapInfos {
+		gatedID := candInfo.SnapID
+		gating := controlled[gatedID]
+		if len(gating) == 0 { // easy case, no refresh control
+			validated = append(validated, candInfo)
+			continue
+		}
+
+		var validationRefs []*asserts.Ref
+
+		fetching := func(f asserts.Fetcher) error {
+			for _, gatingID := range gating {
+				valref := &asserts.Ref{
+					Type:       asserts.ValidationType,
+					PrimaryKey: []string{release.Series, gatingID, gatedID, candInfo.Revision.String()},
+				}
+				err := f.Fetch(valref)
+				if err != nil {
+					return fmt.Errorf("cannot find validation by %q: %v", gatingNames[gatingID], err)
```

Comment:
Would be nice to have a more friendly message for the typical case, which is when the validation was simply not found because it wasn't created by the publisher. Something like `"no validation by %q"`, omitting the stock err.